Update http.Transport if it already exists in ExecProvider

awly opened the pull request:
**What this PR does / why we need it**:
This unbreaks ExecPlugin. Without the change, we hit this error
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/transport/transport.go#L32

**Release note**:
```release-note
Fix kubelet startup failure when using ExecPlugin in kubeconfig
```

awly commented:
/assign @mikedanese 
/cc @liggitt

liggitt commented:
/hold
this reverts the fix in https://github.com/kubernetes/kubernetes/pull/63492 and removes the ability to tear down kubelet connections on heartbeat failure.

awly commented:
@liggitt good point, not sure how i forgot about that :|

Changed to handle existing Transport in ExecProvider and use stdlib `GetCertificate` version.
WDYT?

liggitt commented on staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:
```diff
@@ -179,20 +179,41 @@ func (a *Authenticator) UpdateTransportConfig(c *transport.Config) error {
-			cert, err := getCert()
-			if err != nil {
-				return nil, err
+	if t, ok := c.Transport.(*http.Transport); ok && t != nil {
```
what does this do when run with `--v=6`? might need a `func TransportFor(transport http.RoundTripper) (*http.Transport, error)` helper in pkg/util/net to unwrap nested debugging transports (see `DialerFor` as an example)

awly replied:
Just tested, `--v=6` works fine.
But I agree with your point, someone can introduce a wrapper eventually, breaking this.
Added a `TransportFor` helper.

liggitt commented on staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:
```diff
@@ -179,20 +179,41 @@ func (a *Authenticator) UpdateTransportConfig(c *transport.Config) error {
+		}
+		getCert := t.TLSClientConfig.GetClientCertificate
+		t.TLSClientConfig.GetClientCertificate = func(hi *tls.CertificateRequestInfo) (*tls.Certificate, error) {
+			// If previous GetCert is present and returns a valid non-nil
```
trying to think through the implications if you use client cert rotation *and* an exec plugin... the client cert rotation code would always win. does that make sense?

liggitt replied:
where are we checking if the returned cert is valid?

awly replied:
> where are we checking if the returned cert is valid?

`refreshCredsLocked` does validation. It fetches the actual certificate, validates and caches it. `GetClientCertificate` returns the cached copy.
I added cert expiry validation too in this PR, just in case.

> trying to think through the implications if you use client cert rotation and an exec plugin... the client cert rotation code would always win. does that make sense?

Is cert rotation *and* exec plugin a valid configuration? Does it mean "fetch initial cert via plugin but then rotate manually"?
Our use case relies on exec plugin to bootstrap a fresh cert whenever needed and cert rotation in kubelet is disabled.

liggitt replied:
> refreshCredsLocked does validation.

I was referring to what the previous GetCert was returning ("If previous GetCert is present and returns a *valid* non-nil..."), not what the exec plugin returned

> Is cert rotation and exec plugin a valid configuration?

I don't know, but this is allowing it :)

awly replied:
> I was referring to what the previous GetCert was returning ("If previous GetCert is present and returns a valid non-nil..."), not what the exec plugin returned

Ah, gotcha. I don't think it's this plugin's job to check that. If something returns a certificate here, it better be valid. Otherwise it's a bug.

I'd rather leave it as is, WDYT?

mikedanese replied:
For the sanity of the next person who tries to debug this, I think it would be better to return an error if GetClientCertificate is not nil. That would effectively disallow client cert rotation to be used with the exec plugin, right? I'd be happy to avoid this complexity.

liggitt replied:
> I think it would be better to return an error if GetClientCertificate is not nil. That would effectively disallow client cert rotation to be used with the exec plugin, right? I'd be happy to avoid this complexity.

that seems like a reasonable starting point. mixing the two gets weird.

awly replied:
SGTM, made it an error in both cases

liggitt commented on staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:
```diff
@@ -179,20 +179,41 @@ func (a *Authenticator) UpdateTransportConfig(c *transport.Config) error {
+			if getCert != nil {
+				cert, err := getCert(hi)
+				if err != nil {
+					return nil, err
```
would we want to check the exec plugin in an error case?

awly replied:
Makes sense to me, added.
It's an unlikely configuration, but doesn't hurt.

awly commented:
@liggitt thanks for the comments, PTAL

awly commented:
FYI, I'd like to cherry-pick this into 1.11.2.

mikedanese commented on staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:
```diff
@@ -353,6 +386,19 @@ func (a *Authenticator) refreshCredsLocked(r *clientauthentication.Response) err
+		if certBlock == nil {
+			return errors.New("can't parse client certificate returned by exec plugin as PEM block")
+		}
+		x509Cert, err := x509.ParseCertificate(certBlock.Bytes)
```
Use the cert on line 402 so you don't have to parse the pem twice.

awly replied:
You mean manually parse `ClientKeyData` and build `tls.Certificate` via a literal?
It feels like more work and easy to miss some validations that `tls.X509KeyPair` does

mikedanese reviewed (commented):
The rest client keeps on giving. What if we added a WrapDialer field to the restclient.Config and used that to add the connection tracker? Would that be simpler?

liggitt commented on staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:
```diff
@@ -353,6 +386,19 @@ func (a *Authenticator) refreshCredsLocked(r *clientauthentication.Response) err
+		if err != nil {
+			return fmt.Errorf("failed parsing x509 client certificate returned by exec plugin: %v", err)
+		}
+		if time.Now().After(x509Cert.NotAfter) {
```
are we already checking time client-side elsewhere? this makes us vulnerable to client clock skew

awly replied:
I don't know if we're checking it elsewhere. It should be kube-apiserver's responsibility to verify cert validity on incoming requests. Could you clarify how this makes us vulnerable?

awly commented on staging/src/k8s.io/apimachinery/pkg/util/net/http.go:
```diff
@@ -124,26 +124,40 @@ type RoundTripperWrapper interface {
+// TransportFor returns the root *http.Transport behind all wrapper
+// RoundTrippers. Wrappers must implement RoundTripperWrapper.
+func TransportFor(transport http.RoundTripper) (*http.Transport, error) {
 	if transport == nil {
```
This shifts the nil check from every caller to here. `transport.Config.Transport` can be nil with valid config, afaik (just not in kubelet)

awly commented:
@liggitt @mikedanese PTAL. Turns out we can just set `restclient.Config.Dialer` instead of `Transport`. This fixes the issue in my test cluster.

mikedanese commented on staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:
```diff
@@ -179,21 +182,10 @@ func (a *Authenticator) UpdateTransportConfig(c *transport.Config) error {
-		}
-		return a.cert()
+	if c.TLS.GetCert != nil {
+		return errors.New("can't add TLC certificate callback: transport.Config.TLS.GetCert already set")
```
s/TLS/TLS/?

awly replied:
done

Commit:
Set connrotation dialer via restclient.Config.Dialer

Instead of Transport. This fixes ExecPlugin, which fails if
restclient.Config.Transport is set.

mikedanese commented:
/retest

mikedanese commented:
/retest

liggitt commented:
/lgtm
/hold cancel

marpaia commented:
/sig cli
